src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcServerTransport.cs

```diff
@@ -0,0 +1,163 @@
+                _disposed = true;
+            }
+        }
+
+        public override async Task<Stream> AcceptAsync(CancellationToken token)
+        {
+            NamedPipeServerStream connectedStream;
+            try
+            {
+                await _stream.WaitForConnectionAsync(token);
+
+                connectedStream = _stream;
+            }
+            finally
+            {
+                CreateNewPipeServer();
+            }
+            return connectedStream;
+        }
+
+        private void CreateNewPipeServer()
+        {
+            _stream = new NamedPipeServerStream(_pipeName, PipeDirection.InOut, 10, PipeTransmissionMode.Byte, PipeOptions.Asynchronous);
+        }
+    }
+
+    internal sealed class UnixDomainSocketServerTransport : IpcServerTransport
+    {
+        private bool _disposed = false;
+
+        private readonly string _path;
+        private readonly Socket _socket;
+
+        public UnixDomainSocketServerTransport(string path)
+        {
+            _path = path;
+            _socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
+            _socket.Bind(PidIpcEndpoint.CreateUnixDomainSocketEndPoint(path));
+            _socket.Listen(255);
+            _socket.LingerState.Enabled = false;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                if (disposing)
+                {
+                    try
+                    {
+                        _socket.Shutdown(SocketShutdown.Both);
+                    }
+                    catch { }
+                    finally
+                    {
+                        _socket.Close(0);
+                    }
+                    _socket.Dispose();
+
+                    if (File.Exists(_path))
+                        File.Delete(_path);
+                }
+                _disposed = true;
+            }
+        }
+
+        public override async Task<Stream> AcceptAsync(CancellationToken token)
+        {
+            TaskCompletionSource<Socket> acceptCompletionSource = new TaskCompletionSource<Socket>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            Action cancelAccept = () =>
+            {
+                acceptCompletionSource.TrySetCanceled(token);
+                _socket.Close(0);
+            };
+
+            AsyncCallback endAccept = result =>
+            {
+                if (!token.IsCancellationRequested)
+                {
+                    try
+                    {
+                        acceptCompletionSource.TrySetResult(_socket.EndAccept(result));
+                    }
+                    catch (Exception ex)
+                    {
+                        acceptCompletionSource.TrySetException(ex);
+                    }
+                }
+            };
+
+            using (token.Register(cancelAccept))
+            {
+                _socket.BeginAccept(endAccept, null);
```

Comment:
Can we use AcceptAsync instead? I also believe there are some utility apis that convert IAsyncResult to Task

Reply:
I'll look into that.

Regarding conversion of IAsyncResult into Task, you do that with TaskFactory.FromAsync, however none of those allow for cancellation. I would still have to register for cancellation and close the socket. And even then, the Task would probably be in the faulted state rather than the cancelled state (I could await it here, catch exception if cancellation was requested, and call token.ThrowIfCancellationRequested). I can look into that as an alternative too.